### PR TITLE
 [BOLT] Add sink block to flow CFG in profile inference

### PR DESCRIPTION
**Pull request by shawbyoung**
Summary: Constructing an artificial sink block for the
flow CFG in stale profile inference to allow profile
inference to be run on CFGs with blocks that terminate
and have successors.

Testing Plan: Added infer_no_exits.test to verify that 
functions with exit blocks with a landing pad are 
covered by stale profile inference.

### CODE REVIEW

File: bolt/lib/Profile/StaleProfileMatching.cpp [hunk 1]
```diff
@@ -309,22 +309,33 @@ createFlowFunction(const BinaryFunction::BasicBlockOrderType &BlockOrder) {
   FlowFunction Func;
 
   // Add a special "dummy" source so that there is always a unique entry point.
-  // Because of the extra source, for all other blocks in FlowFunction it holds
-  // that Block.Index == BB->getIndex() + 1
   FlowBlock EntryBlock;
   EntryBlock.Index = 0;
   Func.Blocks.push_back(EntryBlock);
 
+  auto BinaryBlockIsExit = [&](const BinaryBasicBlock &BB) {
```
Comment (aaupov) [inline bolt/lib/Profile/StaleProfileMatching.cpp:316]: Since we're only using it in a single place, can avoid defining this helper lambda and use the underlying check directly.

File: bolt/lib/Profile/StaleProfileMatching.cpp [hunk 2]
```diff
@@ -309,22 +309,33 @@ createFlowFunction(const BinaryFunction::BasicBlockOrderType &BlockOrder) {
...
+      return true;
+    return false;
+  };
+
   // Create FlowBlock for every basic block in the binary function
   for (const BinaryBasicBlock *BB : BlockOrder) {
     Func.Blocks.emplace_back();
     FlowBlock &Block = Func.Blocks.back();
     Block.Index = Func.Blocks.size() - 1;
+    Block.HasSuccessors = BinaryBlockIsExit(*BB);
```
Comment (aaupov) [inline bolt/lib/Profile/StaleProfileMatching.cpp:327]: Let's call the new attribute `IsExit` so the intent is clear.

File: bolt/lib/Profile/StaleProfileMatching.cpp [hunk 3]
```diff
@@ -360,25 +371,37 @@ createFlowFunction(const BinaryFunction::BasicBlockOrderType &BlockOrder) {
   // Add dummy edges to the extra sources. If there are multiple entry blocks,
   // add an unlikely edge from 0 to the subsequent ones
   assert(InDegree[0] == 0 && "dummy entry blocks shouldn't have predecessors");
-  for (uint64_t I = 1; I < Func.Blocks.size(); I++) {
```
Comment (aaupov) [inline bolt/lib/Profile/StaleProfileMatching.cpp:363]: Keep Func.Blocks.size() and add `-1` to skip SinkBlock. Add a comment about skipping it.

File: bolt/lib/Profile/StaleProfileMatching.cpp [hunk 4]
```diff
@@ -360,25 +371,37 @@ createFlowFunction(const BinaryFunction::BasicBlockOrderType &BlockOrder) {
...
+  for (uint64_t I = 1; I < BlockOrder.size() + 1; I++) {
     const BinaryBasicBlock *BB = BlockOrder[I - 1];
     if (BB->isEntryPoint() || InDegree[I] == 0) {
       Func.Jumps.emplace_back();
       FlowJump &Jump = Func.Jumps.back();
-      Jump.Source = 0;
+      Jump.Source = Func.Entry;
```
Comment (aaupov) [inline bolt/lib/Profile/StaleProfileMatching.cpp:379]: We're not changing the index of the entry, right? So this change is not needed.

File: bolt/lib/Profile/StaleProfileMatching.cpp [hunk 5]
```diff
@@ -360,25 +371,37 @@ createFlowFunction(const BinaryFunction::BasicBlockOrderType &BlockOrder) {
...
+  }
+
   // Create necessary metadata for the flow function
   for (FlowJump &Jump : Func.Jumps) {
     assert(Jump.Source < Func.Blocks.size());
     Func.Blocks[Jump.Source].SuccJumps.push_back(&Jump);
     assert(Jump.Target < Func.Blocks.size());
     Func.Blocks[Jump.Target].PredJumps.push_back(&Jump);
   }
+
```
Comment (aaupov) [inline bolt/lib/Profile/StaleProfileMatching.cpp:404]: nit: avoid spurious changes

File: bolt/lib/Profile/StaleProfileMatching.cpp [hunk 6]
```diff
@@ -640,6 +663,9 @@ void assignProfile(BinaryFunction &BF,
       if (Jump->Flow == 0)
         continue;
 
+      // Skip the artificial sink block
```
Comment (aaupov) [inline bolt/lib/Profile/StaleProfileMatching.cpp:666]: Can instead use `BlockOrder.size() - 1` to avoid the check on each block.

File: llvm/include/llvm/Transforms/Utils/SampleProfileInference.h [hunk 7]
```diff
@@ -31,10 +31,10 @@ struct FlowBlock {
   uint64_t Flow{0};
   std::vector<FlowJump *> SuccJumps;
   std::vector<FlowJump *> PredJumps;
+  bool HasSuccessors{false};
 
   /// Check if it is the entry block in the function.
   bool isEntry() const { return PredJumps.empty(); }
-
```
Comment (aaupov) [inline llvm/include/llvm/Transforms/Utils/SampleProfileInference.h:37]: ditto

File: bolt/lib/Profile/StaleProfileMatching.cpp (hunk 6 above)
Comment (shawbyoung) [reply bolt/lib/Profile/StaleProfileMatching.cpp:666]: I'm not seeing where "BlockOrder.size() - 1" can be used. This check "if (Jump->Target == Func.Sink) continue;" ensures that the assertion within the index operation here "BinaryBasicBlock &SuccBB = *BlockOrder[Jump->Target - 1];" doesn't fail

Comment (aaupov) [reply bolt/lib/Profile/StaleProfileMatching.cpp:660]: You're right. Didn't get the full context of this loop.

File: bolt/lib/Profile/StaleProfileMatching.cpp [hunk 8]
```diff
@@ -320,11 +318,19 @@ createFlowFunction(const BinaryFunction::BasicBlockOrderType &BlockOrder) {
     Func.Blocks.emplace_back();
     FlowBlock &Block = Func.Blocks.back();
     Block.Index = Func.Blocks.size() - 1;
+    if (BB->successors().empty())
+      Block.IsExit = true;
```
Comment (aaupov) [inline bolt/lib/Profile/StaleProfileMatching.cpp:322]:
```suggestion
    Block.IsExit = BB->successors().empty();
```

**Comment by aaupov** [comment]
LG with a small nit. Thank you.

File: bolt/lib/Profile/StaleProfileMatching.cpp [hunk 9]
```diff
@@ -372,6 +377,17 @@ createFlowFunction(const BinaryFunction::BasicBlockOrderType &BlockOrder) {
     }
   }
 
+  // Add dummy edges from the exit blocks to the sink block.
+  for (uint64_t I = 1; I < BlockOrder.size() + 1; I++) {
+    FlowBlock &Block = Func.Blocks[I];
```
Comment (ayermolo) [inline bolt/lib/Profile/StaleProfileMatching.cpp:382]: const?

File: bolt/lib/Profile/StaleProfileMatching.cpp [hunk 10]
```diff
@@ -320,11 +318,18 @@ createFlowFunction(const BinaryFunction::BasicBlockOrderType &BlockOrder) {
     Func.Blocks.emplace_back();
     FlowBlock &Block = Func.Blocks.back();
     Block.Index = Func.Blocks.size() - 1;
+    Block.IsExit = BB->successors().empty();
     (void)BB;
     assert(Block.Index == BB->getIndex() + 1 &&
            "incorrectly assigned basic block index");
   }
 
+  // Add a special "dummy" sink block so there is always a unique sink
```
Comment (ayermolo) [inline bolt/lib/Profile/StaleProfileMatching.cpp:327]: nit period

File: bolt/lib/Profile/StaleProfileMatching.cpp [hunk 11]
```diff
@@ -320,11 +318,18 @@ createFlowFunction(const BinaryFunction::BasicBlockOrderType &BlockOrder) {
...
+  FlowBlock SinkBlock;
+  SinkBlock.Index = Func.Blocks.size();
+  Func.Blocks.push_back(SinkBlock);
+  Func.Sink = SinkBlock.Index;
+
   // Create FlowJump for each jump between basic blocks in the binary function
```
Comment (ayermolo) [inline bolt/lib/Profile/StaleProfileMatching.cpp:333]: same

File: bolt/lib/Profile/StaleProfileMatching.cpp [hunk 12]
```diff
@@ -372,6 +377,17 @@ createFlowFunction(const BinaryFunction::BasicBlockOrderType &BlockOrder) {
     }
   }
 
+  // Add dummy edges from the exit blocks to the sink block.
+  for (uint64_t I = 1; I < BlockOrder.size() + 1; I++) {
+    FlowBlock &Block = Func.Blocks[I];
+    if (Block.IsExit) {
```
Comment (maksfb) [inline bolt/lib/Profile/StaleProfileMatching.cpp:383]: Nit: use "early continue" (https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code). Since `Block` is not used anywhere else, I'd just write:
```
  if (!Func.Blocks[I].IsExit)
    continue;

File: llvm/include/llvm/Transforms/Utils/SampleProfileInference.h [hunk 13]
```diff
@@ -57,6 +58,7 @@ struct FlowFunction {
   std::vector<FlowJump> Jumps;
   /// The index of the entry block.
   uint64_t Entry{0};
+  uint64_t Sink{0};
```
Comment (maksfb) [inline llvm/include/llvm/Transforms/Utils/SampleProfileInference.h:61]: Nit: add a description and initialize to `uint64_t` max value to indicate that a proper initialization will happen later.

**Review by maksfb — COMMENTED** [review]
Add a brief description why you are adding the sink block to the summary.

File: bolt/lib/Profile/StaleProfileMatching.cpp [hunk 14]
```diff
@@ -309,23 +309,28 @@ createFlowFunction(const BinaryFunction::BasicBlockOrderType &BlockOrder) {
...
   EntryBlock.Index = 0;
   Func.Blocks.push_back(EntryBlock);
 
-  // Create FlowBlock for every basic block in the binary function
+  // Create FlowBlock for every basic block in the binary function.
   for (const BinaryBasicBlock *BB : BlockOrder) {
     Func.Blocks.emplace_back();
     FlowBlock &Block = Func.Blocks.back();
     Block.Index = Func.Blocks.size() - 1;
+    Block.IsExit = BB->successors().empty();
```
Comment (aaupov) [inline bolt/lib/Profile/StaleProfileMatching.cpp:321]: I thought a bit more about this change and realized it's not NFC for the following reason: we used to determine whether a block is an exit depending on both CFG successors AND landing pads (see line 351 below)
`351     for (const BinaryBasicBlock *DstBB : SrcBB->landing_pads()) {`

In the new IsExit determination, we will only use CFG edges and omit LPs. This omission should help the case we want to target (degenerate graphs with no blocks without CFG/LP successors), but we'll need to add a test for it.

**Review by aaupov — CHANGES_REQUESTED** [review]
Please drop [NFC] from the title + need a test (see comment)

File: bolt/lib/Profile/StaleProfileMatching.cpp [hunk 15]
```diff
@@ -595,9 +595,7 @@ bool canApplyInference(const FlowFunction &Func) {
   if (Func.Blocks.size() > opts::StaleMatchingMaxFuncSize)
     return false;
 
-  bool HasExitBlocks = llvm::any_of(
-      Func.Blocks, [&](const FlowBlock &Block) { return Block.isExit(); });
-  if (!HasExitBlocks)
+  if (Func.Blocks[Func.Sink].isEntry())
```
Comment (aaupov) [inline bolt/lib/Profile/StaleProfileMatching.cpp:598]: Please add a comment what is meant by Sink block returning true for `isEntry`

File: llvm/include/llvm/Transforms/Utils/SampleProfileInference.h [hunk 16]
```diff
@@ -28,6 +28,7 @@ struct FlowBlock {
   uint64_t Weight{0};
   bool HasUnknownWeight{true};
   bool IsUnlikely{false};
+  bool IsExit{false};
```
Comment (WenleiHe) [inline llvm/include/llvm/Transforms/Utils/SampleProfileInference.h:31]: This is going to be confusing as we already have `isExit` based on whether there is outgoing edges. What's different for the boolean, and what's our definition of "is exit block"?

File: bolt/lib/Profile/StaleProfileMatching.cpp [hunk 17]
```diff
@@ -309,23 +309,28 @@ createFlowFunction(const BinaryFunction::BasicBlockOrderType &BlockOrder) {
...
     FlowBlock &Block = Func.Blocks.back();
     Block.Index = Func.Blocks.size() - 1;
+    Block.IsExit = BB->successors().empty();
     (void)BB;
     assert(Block.Index == BB->getIndex() + 1 &&
            "incorrectly assigned basic block index");
   }
 
-  // Create FlowJump for each jump between basic blocks in the binary function
+  // Add a special "dummy" sink block so there is always a unique sink.
```
Comment (WenleiHe) [inline bolt/lib/Profile/StaleProfileMatching.cpp:327]: I notice there is always special entry block, so there is precedence. But I'm still wondering whether this should be handled inside profi instead? Like alternatively, adding artificial entry/sink can be done inside profi as needed, so this is all transparent to the users of profi.

Comment (shawbyoung) [reply bolt/lib/Profile/StaleProfileMatching.cpp:327]: In order to construct the artificial sink block and edges from exit blocks to the sink block, BinaryBasicBlock information is/will be needed, which is not currently exposed to SampleProfileInference/profi. For instance, identifying calls to functions that pass control back outside of the function (i.e. calls to std::terminate) requires BinaryBasicBlock information. (Identifying these calls is something I plan to implement in a subsequent commit). I think it also makes sense organizationally to do this in the construction of the flow function - it would be inelegant to expose BinaryBasicBlock information to profi (code duplicity) just to add the sink block. The construction of the entry block could be more easily moved to profi, but given that sink block should be constructed within createFlowFunction it makes sense for that block's construction to remain where it is as well.

File: bolt/test/X86/Inputs/infer_no_exits.s [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+
+	.hidden	DW.ref.__gxx_personality_v0
+	.weak	DW.ref.__gxx_personality_v0
+	.section	.data.DW.ref.__gxx_personality_v0,"awG",@progbits,DW.ref.__gxx_personality_v0,comdat
+	.p2align	3, 0x0
+	.type	DW.ref.__gxx_personality_v0,@object
+	.size	DW.ref.__gxx_personality_v0, 8
+DW.ref.__gxx_personality_v0:
+	.quad	__gxx_personality_v0
+	.ident	"Facebook clang version 19.0.0 (https://git.internal.tfbnw.net/repos/git/rw/osmeta/external/llvm-project 4fc35430a0be5afb4ff7909ef6f51859b7019e33)"
```
Comment (aaupov) [inline bolt/test/X86/Inputs/infer_no_exits.s:168]: Let's drop that

File: bolt/test/X86/infer_no_exits.test [hunk 19]
```diff
@@ -1,8 +1,10 @@
 ## This verifies that functions where an exit block has a landing pad are covered by stale profile inference.
-# RUN: %clangxx %cxxflags %p/Inputs/infer_no_exits.cpp -o %t.exe
+# RUN: %clangxx %cxxflags %p/Inputs/infer_no_exits.s -o %t.exe
 # RUN: link_fdata %s %t.exe %t.preagg PREAGG
 # RUN: perf2bolt %t.exe -p %t.preagg --pa -o %t.fdata -w %t.yaml
 # RUN: sed -i 's/0000/0001/g' %t.yaml
```
Comment (aaupov) [inline bolt/test/X86/infer_no_exits.test:5]: Don't need this line

File: bolt/test/X86/infer_no_exits.test [hunk 20]
```diff
@@ -1,8 +1,10 @@
...
+# RUN: sed -i 's/hash:            0x9AE3FDBFFBE36274/hash:            0x0000000000000000/g' %t.yaml
```
Comment (aaupov) [inline bolt/test/X86/infer_no_exits.test:6]: Let's make it robust wrt possible BF hash changes: match the line with BF hash and replace the hash with 0x0000000000000000.

File: bolt/lib/Profile/StaleProfileMatching.cpp (hunk 17 above)
Comment (WenleiHe) [reply bolt/lib/Profile/StaleProfileMatching.cpp:327]: I don't see why do we need BinaryBasicBlock info in profi. So we construct a CFG represented with flow blocks first. BinaryBasicBlock is only used during CFG construction like it is now. Then within profi, if we find multiple exit blocks (no successors) in that CFG, we just connect these exit blocks into a dummy sink block, without needing info from original BinaryBasicBlock.

File: llvm/include/llvm/Transforms/Utils/SampleProfileInference.h (hunk 16 above)
Comment (shawbyoung) [reply llvm/include/llvm/Transforms/Utils/SampleProfileInference.h]: I agree, just changed this representation

File: bolt/lib/Profile/StaleProfileMatching.cpp (hunk 17 above)
Comment (shawbyoung) [reply bolt/lib/Profile/StaleProfileMatching.cpp:329]: We would need binary block info because the condition "block has no successors" does not fully encompass our definition of "exit blocks". For instance, there are landing pads flow blocks that have cfg successors but call/transitively std::terminate or some other function that transfers control back outside of the function, thus qualifying it as an exit block. We need binary block information to parse through instructions and identify these calls. I plan to cover these degenerate cases in a following commit.

File: llvm/include/llvm/Transforms/Utils/SampleProfileInference.h [hunk 21]
```diff
@@ -57,6 +57,7 @@ struct FlowFunction {
   std::vector<FlowJump> Jumps;
   /// The index of the entry block.
   uint64_t Entry{0};
+  uint64_t Sink{UINT64_MAX};
```
Comment (aaupov) [inline llvm/include/llvm/Transforms/Utils/SampleProfileInference.h:60]: Let's also keep track of the Sink block on BOLT's side

File: bolt/lib/Profile/StaleProfileMatching.cpp (hunk 17 above)
Comment (WenleiHe) [reply bolt/lib/Profile/StaleProfileMatching.cpp:330]: > "block has no successors" does not fully encompass our definition of "exit blocks"

What is our definition of "exit blocks"? A reasonable definition could be a block that transfer control outside of the function. Now successor blocks naturally implies the successor within the same function. With that definition, blocks containing no return calls should not have successors, and such information should be embeded in the CFG already (`isExitBlock`), without needing to look at BinaryBasicBlock again. If we do need extra info to know where we have exit flow due to noreturn call after CFG is constructed, that indicates the CFG construction is wrong? 

> I plan to cover these degenerate cases in a following commit. 

Evidently, the way this changes are split causes confusion. This patch as it is doesn't make its intention obvious. It would be better to just include those intended changes in one patch.

File: bolt/lib/Profile/StaleProfileMatching.cpp [hunk 22]
```diff
@@ -307,25 +307,32 @@ void BinaryFunction::computeBlockHashes(HashFunction HashFunction) const {
 FlowFunction
 createFlowFunction(const BinaryFunction::BasicBlockOrderType &BlockOrder) {
   FlowFunction Func;
+  std::vector<uint64_t> ExitBlocksIndices;
```
Comment (WenleiHe) [inline bolt/lib/Profile/StaleProfileMatching.cpp:310]: Again, we have to have a consistent definition of what is an exit block. It doesn't matter where you put it, another flag in `FlowBlock` or extra bookkeeping here. Why do we need to have extra bookkeeping given FlowBlock already has `isExitBlock`.

**Review by WenleiHe — CHANGES_REQUESTED** [review]

File: bolt/lib/Profile/StaleProfileMatching.cpp (hunk 17 above)
Comment (shawbyoung) [reply bolt/lib/Profile/StaleProfileMatching.cpp:326]: > With that definition, blocks containing no return calls should not have successors,
I don’t think this is universally true for exceptions. Consider a block that has a return instruction and throws an exception. The BBB representation would report that there are no successors, but in our current construction of the flow CFG, we would not label that as an exit block – the landing pad would be considered a successor and we’d lose out on applying profile inference to that function. However, this new implementation “ignores” landing pads in the identification of exit blocks which increase our scope (the added test reflects this). I took your advice on IsExit and ExitBlockIndicies – those were redundant as you pointed out and sorry about the confusion.

File: bolt/lib/Profile/StaleProfileMatching.cpp [hunk 23]
```diff
@@ -342,6 +345,15 @@ createFlowFunction(const BinaryFunction::BasicBlockOrderType &BlockOrder) {
       InDegree[Jump.Target]++;
       UniqueSuccs.insert(DstBB);
     }
+
+    // If the block is an exit, add a dummy edge from it to the sink block.
+    if (UniqueSuccs.empty()) {
+      Func.Jumps.emplace_back();
+      FlowJump &Jump = Func.Jumps.back();
+      Jump.Source = SrcBB->getIndex() + 1;
+      Jump.Target = Func.Blocks.size() - 1;
```
Comment (aaupov) [inline bolt/lib/Profile/StaleProfileMatching.cpp:354]: Do we not need InDegree?

File: bolt/test/X86/infer_no_exits.test [hunk 24]
```diff
@@ -1,5 +1,4 @@
 ## This verifies that functions where an exit block has a landing pad are covered by stale profile inference.
-# RUN: %clangxx %cxxflags %p/Inputs/infer_no_exits.s -o %t.exe
```
Comment (aaupov) [inline bolt/test/X86/infer_no_exits.test]: Wait, why did you remove that?

File: bolt/test/X86/infer_no_exits.test (hunk 19 above)
Comment (aaupov) [reply bolt/test/X86/infer_no_exits.test:4]: Didn't remove the line

File: bolt/test/X86/infer_no_exits.test [hunk 25]
```diff
@@ -0,0 +1,13 @@
...
+# RUN: link_fdata %s %t.exe %t.preagg PREAGG
+# RUN: perf2bolt %t.exe -p %t.preagg --pa -o %t.fdata -w %t.yaml
+# RUN: sed -i 's/0000/0001/g' %t.yaml
+# RUN: sed -i '0,/hash:/s/0x[0-9A-Fa-f]\{16\}/0x0000000000000000/' %t.yaml
+# RUN: llvm-bolt %t.exe -data %t.yaml -o %t.null -v=1 -infer-stale-profile 2>&1 \
+# RUN:   | FileCheck %s
+
+# PREAGG: B X:0 #main# 1 0
+
+# CHECK: BOLT-INFO
```
Comment (aaupov) [inline bolt/test/X86/infer_no_exits.test:11]: Please clean up this line.

**Review by aaupov — APPROVED** [review]
This state looks good to me. @WenleiHe - do you have any unresolved questions or concerns? We'd like to keep the FlowGraph construction logic in BOLT for now.

File: bolt/lib/Profile/StaleProfileMatching.cpp (hunk 17 above)
Comment (WenleiHe) [reply bolt/lib/Profile/StaleProfileMatching.cpp:326]: Had an offline discussion with @shawbyoung, now I see the problem (thanks for explaining this with actual CFG). So a real exit block can have exception, which makes landing pad its successor, and since landing can recover from exception and continue execution, it will go back to the function. Based on our discussions, there are a few things we need to do:

1) this issue shouldn't be unique to BOLT, worth checking how compiler is handling it. anytime we do something specific for compiler or bolt, if feels fishy because their need should be similar and it should be handled in infer without duplication.

2) we actually need to tell profi that exception flow is rare and it should take that into consideration when doing inference. one simple approximation is just to remove edges to landing pad altogether so profi won't send flow to exception path which is supposed to be very very rare anyways. doing so also solves the problem without adding sink node. However it looks like from the compiler side, we communicate such constraints via "unlikely" flag on edge, so profi can adjust edge costs. So we could also do the same for bolt. One way or another we need to profi to take it into consideration. 

It's okay if we leverage unlikely flag for EH path in a follow up patch, but we probably want an answer for #1 here.

Comment (shawbyoung) [reply bolt/lib/Profile/StaleProfileMatching.cpp:326]: The case handled by this PR (exit block, e.g. a block that ends with return/tail call, containing an exception throwing function call) is unique to BOLT. The compiler maintains the invariant "Invoke ends the block” while BOLT treats Invokes as regular calls that don't terminate the block, and merges the fallthrough block which might be an exit block. Work should be done to mirror this in BOLT - splitting blocks with exception throwing calls at the call and adding the landing pad as the successor. However, this change would be extensive, and this PR is a relatively speaking minimally invasive change that is functionally equivalent.

Comment (WenleiHe) [reply bolt/lib/Profile/StaleProfileMatching.cpp:326]: Sounds good. Agreed that splitting the flow from invoke is more natural. 

Can you add a comment document this as a TODO, so we know why there is divergence in how we handle flow to landing pad? Otherwise this patch looks good. Also helpful to have a TODO comment for unlikely jump flag to communicate to profi. 

Thanks for working through the feedbacks.

**Review by WenleiHe — APPROVED** [review]
lgtm, thanks!
